// rc-scroll-anim/lib/ScrollLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _tweenFunctions = require('tween-functions');

var _tweenFunctions2 = _interopRequireDefault(_tweenFunctions);

var _raf = require('raf');

var _raf2 = _interopRequireDefault(_raf);

var _EventDispatcher = require('./EventDispatcher');

var _EventDispatcher2 = _interopRequireDefault(_EventDispatcher);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Created by jljsj on 16/1/13.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


function noop() {}

var scrollLinkLists = [];

var ScrollLink = function (_React$Component) {
  _inherits(ScrollLink, _React$Component);

  function ScrollLink() {
    _classCallCheck(this, ScrollLink);

    var _this = _possibleConstructorReturn(this, _React$Component.apply(this, arguments));

    _this.onClick = function (e) {
      e.preventDefault();
      if (_this.rafID !== -1) {
        return;
      }
      var docRect = document.documentElement.getBoundingClientRect();
      var elementDom = document.getElementById(_this.props.to);
      var elementRect = elementDom.getBoundingClientRect();
      _this.scrollTop = (0, _util.currentScrollTop)();
      var toTop = Math.round(elementRect.top) - Math.round(docRect.top) - _this.props.offsetTop;
      var t = (0, _util.transformArguments)(_this.props.showHeightActive)[0];
      var toShow = t.match('%') ? _this.clientHeight * parseFloat(t) / 100 : t;
      _this.toTop = _this.props.toShowHeight ? toTop - toShow + 0.5 : toTop;
      _this.initTime = Date.now();
      _this.rafID = (0, _raf2["default"])(_this.raf);
      _EventDispatcher2["default"].removeAllType('scroll.scrollAnchorEvent');
      scrollLinkLists.forEach(function (item) {
        if (item !== _this) {
          item.remActive();
        }
      });
      _this.addActive();
    };

    _this.addScrollEventListener = function () {
      var date = Date.now();
      var length = _EventDispatcher2["default"]._listeners.scroll ? _EventDispatcher2["default"]._listeners.scroll.length : 0;
      _this.eventType = 'scroll.scrollAnchorEvent' + date + length;
      _EventDispatcher2["default"].addEventListener(_this.eventType, _this.scrollEventListener);
      // 第一次进入；等全部渲染完成后执行;
      setTimeout(function () {
        _this.scrollEventListener();
      });
    };

    _this.raf = function () {
      if (_this.rafID === -1) {
        return;
      }
      var duration = _this.props.duration;
      var now = Date.now();
      var progressTime = now - _this.initTime > duration ? duration : now - _this.initTime;
      var easeValue = _tweenFunctions2["default"][_this.props.ease](progressTime, _this.scrollTop, _this.toTop, duration);
      window.scrollTo(window.scrollX, easeValue);
      if (progressTime === duration) {
        _this.cancelRequestAnimationFrame();
        _EventDispatcher2["default"].reAllType('scroll.scrollAnchorEvent');
      } else {
        _this.rafID = (0, _raf2["default"])(_this.raf);
      }
    };

    _this.cancelRequestAnimationFrame = function () {
      _raf2["default"].cancel(_this.rafID);
      _this.rafID = -1;
    };

    _this.addActive = function () {
      if (!_this.state.active) {
        var obj = {
          target: _this.dom,
          to: _this.props.to
        };
        _this.props.onFocus(obj);
        _this.setState({
          active: true
        }, function () {
          if (_this.props.toHash) {
            var link = '#' + _this.props.to;
            history.pushState(null, window.title, link);
          }
        });
      }
    };

    _this.remActive = function () {
      if (_this.state.active) {
        var obj = {
          target: _this.dom,
          to: _this.props.to
        };
        _this.props.onBlur(obj);
        _this.setState({
          active: false
        });
      }
    };

    _this.scrollEventListener = function () {
      var docRect = document.documentElement.getBoundingClientRect();
      _this.clientHeight = (0, _util.windowHeight)();
      var elementDom = document.getElementById(_this.props.to);
      if (!elementDom) {
        throw new Error('There is no to(' + _this.props.to + ') in the element.');
      }
      var elementRect = elementDom.getBoundingClientRect();
      var elementClientHeight = elementDom.clientHeight;
      var scrollTop = (0, _util.currentScrollTop)();
      var top = Math.round(docRect.top - elementRect.top + scrollTop);
      var showHeightActive = (0, _util.transformArguments)(_this.props.showHeightActive);
      var startShowHeight = showHeightActive[0].toString().indexOf('%') >= 0 ? parseFloat(showHeightActive[0]) / 100 * _this.clientHeight : parseFloat(showHeightActive[0]);
      var endShowHeight = showHeightActive[1].toString().indexOf('%') >= 0 ? parseFloat(showHeightActive[1]) / 100 * _this.clientHeight : parseFloat(showHeightActive[1]);
      if (top >= -startShowHeight && top < elementClientHeight - endShowHeight) {
        _this.addActive();
      } else {
        _this.remActive();
      }
    };

    _this.rafID = -1;
    _this.state = {
      active: false
    };
    if (_this.props.location) {
      throw new Error('ScrollLink "location" was abandoned, please use "to"');
    }
    return _this;
  }

  ScrollLink.prototype.componentDidMount = function componentDidMount() {
    this.dom = _reactDom2["default"].findDOMNode(this);
    scrollLinkLists.push(this);
    if (this.props.onAsynchronousAddEvent) {
      this.props.onAsynchronousAddEvent(this.addScrollEventListener);
    } else {
      this.addScrollEventListener();
    }
  };

  ScrollLink.prototype.componentWillUnmount = function componentWillUnmount() {
    var _this2 = this;

    scrollLinkLists = scrollLinkLists.filter(function (item) {
      return item !== _this2;
    });
    _EventDispatcher2["default"].removeEventListener(this.eventType, this.scrollEventListener);
    this.cancelRequestAnimationFrame();
  };

  ScrollLink.prototype.render = function render() {
    var _this3 = this;

    var active = this.state.active ? this.props.active : '';
    var _onClick = this.props.onClick;
    var props = _extends({}, this.props, {
      onClick: function onClick(e) {
        _onClick(e);
        _this3.onClick(e);
      }
    });
    ['component', 'duration', 'active', 'location', 'showHeightActive', 'ease', 'toShowHeight', 'offsetTop', 'to', 'onAsynchronousAddEvent', 'toHash'].forEach(function (key) {
      return delete props[key];
    });
    var reg = new RegExp(active, 'ig');
    var className = props.className || '';
    props.className = className.indexOf(active) === -1 ? (className + ' ' + active).trim() : className.replace(reg, '').trim();
    return (0, _react.createElement)(this.props.component, props);
  };

  return ScrollLink;
}(_react2["default"].Component);

ScrollLink.propTypes = {
  component: _react2["default"].PropTypes.string,
  children: _react2["default"].PropTypes.any,
  className: _react2["default"].PropTypes.string,
  style: _react2["default"].PropTypes.any,
  offsetTop: _react2["default"].PropTypes.number,
  duration: _react2["default"].PropTypes.number,
  active: _react2["default"].PropTypes.string,
  location: _react2["default"].PropTypes.string,
  to: _react2["default"].PropTypes.string,
  showHeightActive: _react2["default"].PropTypes.any,
  toShowHeight: _react2["default"].PropTypes.bool,
  ease: _react2["default"].PropTypes.string,
  onClick: _react2["default"].PropTypes.func,
  onFocus: _react2["default"].PropTypes.func,
  onBlur: _react2["default"].PropTypes.func,
  onAsynchronousAddEvent: _react2["default"].PropTypes.func,
  toHash: _react2["default"].PropTypes.bool
};

ScrollLink.defaultProps = {
  component: 'div',
  offsetTop: 0,
  duration: 450,
  active: 'active',
  showHeightActive: '50%',
  ease: 'easeInOutQuad',
  toHash: true,
  onClick: noop,
  onFocus: noop,
  onBlur: noop
};

exports["default"] = ScrollLink;
module.exports = exports['default'];
